opal/mca/accelerator/accelerator.h

```diff
@@ -0,0 +1,480 @@
+ */
+typedef enum {
+    MCA_ACCELERATOR_TRANSFER_UNSPEC = 0,
+    MCA_ACCELERATOR_TRANSFER_HTOH,
+    MCA_ACCELERATOR_TRANSFER_HTOD,
+    MCA_ACCELERATOR_TRANSFER_DTOH,
+    MCA_ACCELERATOR_TRANSFER_DTOD,
+} opal_accelerator_transfer_type_t;
+
+struct opal_accelerator_stream_t {
+    opal_object_t super;
+    /* Stream object */
+    void *stream;
+};
+typedef struct opal_accelerator_stream_t opal_accelerator_stream_t;
+OBJ_CLASS_DECLARATION(opal_accelerator_stream_t);
+
+struct opal_accelerator_event_t {
+    opal_object_t super;
+    /* Event object */
+    void *event;
+};
+typedef struct opal_accelerator_event_t opal_accelerator_event_t;
+OBJ_CLASS_DECLARATION(opal_accelerator_event_t);
+
+/**
+ * Check whether a pointer belongs to an accelerator or not.
+ * interfaces
+ *
+ * @param[IN] addr           Pointer to check
+ * @param[OUT] dev_id        Returns the device id against which the memory was allocated
+ * @param[OUT] flags         May be set to indicate additional information
+ *                           about the corresponding pointer such
+ *                           as whether it belongs to unified memory.
+ *
+ * @retval <0                An error has occurred.
+ * @retval 0                 The buffer does not belong to a managed buffer
+ *                           in device memory.
+ * @retval >0                The buffer belongs to a managed buffer in
+ *                           device memory.
+ */
+typedef int (*opal_accelerator_base_module_check_addr_fn_t)(
+    const void *addr, int *dev_id, uint64_t *flags);
+
+/**
+ * Creates a stream for asynchonous operations. This will also
+ * allocate the stream object.
+ *
+ * @param[IN] dev_id         Associated device for the stream
+ * @param[OUT] stream        Returned stream object
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_create_stream_fn_t)(
+    int dev_id, opal_accelerator_stream_t **stream);
+
+/**
+ * Destroys a stream created by opal_accelerator_base_module_create_stream_fn_t().
+ * This will also free the stream object.
+ *
+ * @param[IN] stream         Stream to destroy
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_destroy_stream_fn_t)(
+    opal_accelerator_stream_t *stream);
+
+/**
+ * Waits until all events on the stream are completed.
+ *
+ * @param[IN] stream         Stream to wait for
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_synchronize_stream_fn_t)(
+    opal_accelerator_stream_t *stream);
+
+/**
+ * Creates an event. You must allocate memory for event prior to
+ * this call.
+ *
+ * @param[IN] event          Event to create
+ *
+ * @return                   OPAL_SUCCESS or error status on failure.
+ */
+typedef int (*opal_accelerator_base_module_create_event_fn_t)(
+    opal_accelerator_event_t *event);
```

Comment:
I don't understand why `create_stream` takes a `opal_accelerator_stream_t **stream` and `create_event` takes an `opal_accelerator_event_t *event`.  This implies that the Accelerator API will fully allocate a stream, but require an event to be partially pre-allocated (the create_stream API still allocates).  This causes the user to treat them differently, when they are fundamentally the same thing (a structure with 1 void pointer inside):

```
    opal_accelerator_stream_t *stream = NULL;
    opal_accelerator.create_stream(dev_id, &stream)
    
    opal_accelerator_event_t event = {};
    opal_accelerator.create_event(&event))
```

Reply:
Yeah I might change create_stream to not allocate. The reason I changed create_event to not allocate is because of the asynchronous progress engine. The design wanted to allocate a large block for a circular event array which wouldn't have been possible with create_event doing the allocation.